Prepare publication of the nb-javac to maven central

matthiasblaesing opened the pull request:
- modified build script to include publishing to sonatype OSSRH snapshot
  repository and to the OSSRH staging repository for maven central
- added documentation about the publication process

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

matthiasblaesing committed:
Prepare publication of the nb-javac to maven central

- modified build script to include publishing to sonatype OSSRH snapshot
  repository and to the OSSRH staging repository for maven central
- added documentation about the publication process

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

matthiasblaesing commented:
This is a followup/part of a fix for [NETBEANS-5109](https://issues.apache.org/jira/browse/NETBEANS-5109). I implemented maven central publication for JNA a long time ago and reused parts of that procedure to make it possible to publish nb-javac to maven central.

This PR alone is not enough - the nb-javac team will still need to adjust the pom to create the right person information and update the groupId in the template pom, get accounts from sonatype and do the publishing, but it is the part I can contribute.

I updated the documentation to contain description about the process and I'm willing to give some more info about the process where required.

I used my personal group id to create a snapshot (no I won't publish javac) to demonstrate the result of this PR:

https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/

As you can see, the "-impl.jar" seems to be missing, but that is not the case, maven requires a base artifact and so the implementation became https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/nb-javac-15-20201218.220825-1.jar. 

The "-api.jar" is bundled with a different classifier: https://oss.sonatype.org/content/repositories/snapshots/eu/doppel-helix/com/github/oracle/nb-javac/nb-javac/15-SNAPSHOT/nb-javac-15-20201218.220825-1-api.jar

The javadoc, that is present on maven central is only a place holder, to satisfy the requirements of sonatype.

I thought about creating two artifacts, but I think this is the cleaner solution as this is more or less a direct transfer from the current publication details.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/build.xml:
```diff
@@ -117,26 +117,106 @@
+    </target>
+
+    <!--
+        Publishing ot maven central / OSSRH requires an account from sonatype.
```
typo: `ot`.

matthiasblaesing replied:
Fixed in place.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/pom-nb-javac.xml:
```diff
@@ -0,0 +1,41 @@
+  http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>eu.doppel-helix.com.github.oracle.nb-javac</groupId>
```
Could this line be replaced with a `TEMPLATE` and processed just like the `<version/>` tag in the build script? The publishing script would then have to be invoked with `ant -DgroupId=eu.doppel-helix.com.github.oracle.nb-javac` (or some other group ID). As a benefit, the whole publishing process would be exactly recorded in the repository and anyone with proper OOSRH account could replay it.

matthiasblaesing replied:
Please see my comment on the developer section.

JaroslavTulach replied:
> ... I don't intent to make the groupId a variable. ... the developers will be pretty constant 
> and the same is true for the group id.

I see. Then your PR needs to be modified before it gets to `origin/master`. Can I push my preferred changes into this PR or shall I create my own and use bce95d8 as base?

matthiasblaesing replied:
Sure - this was intended to get the process started and demonstrate, that it is possible with too much fuzz. Feel free to improve, it's appreciated.

JaroslavTulach replied:
There is [PR #1](https://github.com/matthiasblaesing/nb-javac/pull/1) with my changes to make `groupId` pluggable.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/pom-nb-javac.xml:
```diff
@@ -0,0 +1,41 @@
+      <url>https://github.com/matthiasblaesing/</url>
+      <roles>
+        <role>Developer</role>
+      </roles>
```
Congratulation on becoming `nb-javac` developer!

I assume your entry should stay as you have contributed the maven deployment script. In addition to that contacts to @Akshay-Gupta-Oracle, @singh-akhilesh or some generic contact to maintainers of this repository shall be added as well.

matthiasblaesing replied:
No I should not. I won't publish nb-javac, but being listed in the POM is a requirement to be able to publish to maven central. I tested publishing to the snapshot repository and the staging repository and for that to work, I had to be listed. Just as the groupId is currently not usable for the oracle team, this needs to be updated accordingly. 

That is also the reason, that I don't intent to make the groupId a variable.  The pom-nb-javac.xml is a configuration file per group that redistributes. The only thing that regularly changes over time is the version, the developers will be pretty constant and the same is true for the group id.

JaroslavTulach replied:
According to the first paragraph the `nb-javac` maintainers are supposed to __replace__ the developer section with their own information. OK, let them do it.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/build.xml:
```diff
@@ -117,26 +117,106 @@
+
+        With the credentials configured, running
+
+        ant publish-to-ossrh-snapshots
```
If the `groupId` is turned into a template, this line would become:
```
ant publish-to-ossrh-snapshots -DgroupId=your.group.id
```
and the line below would be
```
ant publish-to-maven-central -DgroupId=your.group.id
```

JaroslavTulach reviewed (changes requested):
Thanks a lot Matthias for resolving the technical problems of uploading `nb-javac` properly to Maven central. Once your PR is merged, it would be just a matter of getting the account and handling the upload. I'd prefer if the `groupId` wan't hardcoded in the sources to make it clear the upload is open to everyone.

JaroslavTulach commented:
#1 got merged and now there are conflicts in the `README.md`

matthiasblaesing commented:
I rebased my change onto current master. To get this further, action from the oracle team is needed:

- a groupId needs to be acquired (I don't know the agreements between Oracle and Sonatype or Oracle internal, so I can't really help there) (=> https://github.com/oracle/nb-javac/pull/2/files#diff-4567d841ae351146e993ae3fa61d189ef40a65e7d69da8d31c83720b9f5163e0R7)
- the first set of developers that will do the publishing need to be determined (=> https://github.com/oracle/nb-javac/pull/2/files#diff-4567d841ae351146e993ae3fa61d189ef40a65e7d69da8d31c83720b9f5163e0R30-R39)

With the above to infos we (I can do that) need to update the `pom-nb-javac.xml` to its final contents. After the pom is updated and this PR is merged, a first maven central publication is possible.

Commit:
Make groupId pluggable

Commit:
maven.version shall be just internal variable derived from nb-javac-ver

matthiasblaesing committed:
Update pom.xml with the current main contributors

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

matthiasblaesing committed:
Ensure properties are loaded when the source zip is created

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

matthiasblaesing commented:
@JaroslavTulach Thank you for the update - I merged your changes into this PR. I also updated the pom template to hold the current main contributors and fixed an issue when the source zip was created independently.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/pom-nb-javac.xml:
```diff
@@ -0,0 +1,49 @@
+    </developer>
+    <developer>
+      <email>akshay.ag.gupta@oracle.com</email>
+      <name>Akshay Guptah</name>
```
The `h` at the end is probably a typo.

JaroslavTulach reviewed (approved):
Amazing. This PR seems to have converged to a solution that fully reflects existing reality and allows independent parties to perform the upload. I support merging and seeking for such party to release `nb-javac-15.1`.

matthiasblaesing committed:
Fix typo

Signed-of-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

matthiasblaesing commented:
Thank you for the review and contributions @JaroslavTulach. I fixed the typo and I think is ready to merge @Akshay-Gupta-Oracle

JaroslavTulach commented:
So am I. @lahodaj, would you agree?

lahodaj reviewed (approved):
Looks reasonable to me.

singh-akhilesh reviewed (approved):
:thumbsup:

arvindaprameya commented:
Approved, Thanks for this contribution Matthias .

matthiasblaesing commented:
Thank you all for review. This needs a commiter (I'm not) and then someone to get a group id and get this published. Once that has happend I'm willing to prepare a PR against netbeans to move to the maven central artifact.

Akshay-Gupta-19 reviewed (approved):
Looks correct.

JaroslavTulach commented on make/langtools/netbeans/nb-javac/pom-nb-javac.xml:
```diff
@@ -0,0 +1,49 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
@matthiasblaesing , this file doesn't have proper license header file and gets included on Maven central. Don't you want to create a PR to add the license?
